### PR TITLE
Add predefined HOST ALL memspace

### PR DESCRIPTION
**Pull request by kswiecicki**
- [x] Add tests
- [x] Add documentation

### CODE REVIEW

File: src/memspaces/memspace_numa.h [hunk 1]
```diff
@@ -27,6 +27,8 @@ extern "C" {
 umf_result_t umfMemspaceCreateFromNumaArray(size_t *nodeIds, size_t numIds,
                                             umf_memspace_handle_t *hMemspace);
 
+extern umf_memspace_handle_t UMF_MEMSPACE_HOST;
```
Comment (igchor) [inline src/memspaces/memspace_numa.h:30]: Hm, I think for memspaces we'll need to follow the same approach as with ops - i.e. create a function that returns memspace handle instead of defining a global variable. See: https://github.com/oneapi-src/unified-memory-framework/pull/201

@vinser52 @bratpiorka @pbalcer are we OK with having predefined memspaces as functions? I'm just wondering if we should have a function per memspace or just a single function that accepts some enum (.e.g. `umfMemspaceGet(UMF_MEMSPACE_HOST)`)?

File: src/memspaces/memspace_host.c [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+
+umf_result_t umfMemspaceHostCreate(umf_memspace_handle_t *hMemspace) {
+    if (!hMemspace) {
+        return UMF_RESULT_ERROR_INVALID_ARGUMENT;
+    }
+
+    umf_result_t umf_ret = UMF_RESULT_SUCCESS;
+
+    hwloc_topology_t topology;
+    if (hwloc_topology_init(&topology)) {
```
Comment (igchor) [inline src/memspaces/memspace_host.c:27]: @pbalcer I know we wanted to strip libumf from malloc symbols (hence using base_alloc for internal allocations) but we have not control over what hwloc calls internally (I think they do call malloc). Will linking with hwloc be a problem for the 'proxy library'?

File: src/libumf_macosx.c [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+ *
+ */
+
+#include "memspace_host_internal.h"
+#include "memspace_internal.h"
+#include "provider_tracking.h"
+
+#include <stddef.h>
+
+umf_memory_tracker_handle_t TRACKER = NULL;
```
Comment (igchor) [inline src/libumf_macosx.c:16]: If we really need a file for macosx I'm starting to think that it would be better to have all this initialization logic in a single file with ifdefs vs having 3 separate files for linux, windows and macosx.

Comment (kswiecicki) [reply src/libumf_macosx.c:16]: It was either a new file or checking for some compile option `MACOSX`, because we don't build the OS provider that the HOST memspace depends on.

File: src/memspaces/memspace_host.c (hunk 2 above)
Comment (bratpiorka) [reply src/memspaces/memspace_host.c:27]: I checked that they call malloc (https://github.com/open-mpi/hwloc/blob/eeeda4ad2d0511a2334a02141dab6138b1e0ce47/hwloc/topology.c). If we want to use HOST memspace in the proxy lib that exports malloc symbol, then we'll go into a recursion here

File: src/memspaces/memspace_host.c [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+        umf_ret = UMF_RESULT_ERROR_UNKNOWN;
+        goto err_topology_destroy;
+    }
+
+    // Shouldn't return -1, as 'HWLOC_OBJ_NUMANODE' deosn't appear to be an
+    // object that can be present on multiple levels.
+    // Source: https://www.open-mpi.org/projects/hwloc/doc/hwloc-v2.10.0-letter.pdf
+    size_t nNodes =
+        (size_t)hwloc_get_nbobjs_by_type(topology, HWLOC_OBJ_NUMANODE);
+    size_t *nodeIds = malloc(nNodes * sizeof(size_t));
```
Comment (bratpiorka) [inline src/memspaces/memspace_host.c:42]: please use ba_alloc here - check how this is done for pools

**Review by bratpiorka — CHANGES_REQUESTED** [review]

File: src/libumf_linux.c [hunk 5]
```diff
@@ -7,17 +7,30 @@
...
  */
 
+#include "memspace_host_internal.h"
+#include "memspace_internal.h"
 #include "provider_tracking.h"
 
 #include <stddef.h>
 
 umf_memory_tracker_handle_t TRACKER = NULL;
+umf_memspace_handle_t UMF_MEMSPACE_HOST = NULL;
```
[the remark this answers is not in the document]
Comment (kswiecicki) [reply src/libumf_linux.c:17]: I'm thinking about how to free a memspace initialized that way. Would it be appropriate to free it in library destructor?

File: src/provider/provider_tracking_linux.c [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+ * SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+ *
+ */
+
+#include "provider_tracking.h"
+
+#include <stddef.h>
+
+extern umf_memory_tracker_handle_t TRACKER;
+
```
Comment (igchor) [inline src/provider/provider_tracking_linux.c:15]: Why is this file added here?

File: src/libumf_linux.c (hunk 5 above)
[the remark this answers is not in the document]
Comment (bratpiorka) [reply src/libumf_linux.c:17]: Yes - these pre-defined namespaces are global, singular, and immutable. They can't be destroyed by the user.

[the remark this answers is not in the document]
Comment (kswiecicki) [reply src/libumf_linux.c]: Done.

File: src/memspaces/memspace_host.c [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+
+static bool UMF_MEMSPACE_HOST_INITIALIZED = false;
+static umf_memspace_handle_t UMF_MEMSPACE_HOST = NULL;
+
+umf_result_t umfMemspaceHostGet(umf_memspace_handle_t *hMemspace) {
+    if (!hMemspace) {
+        return UMF_RESULT_ERROR_INVALID_ARGUMENT;
+    }
+
+    if (!UMF_MEMSPACE_HOST_INITIALIZED) {
```
Comment (bratpiorka) [inline src/memspaces/memspace_host.c:83]: can't you just check if UMF_MEMSPACE_HOST == NULL ?

**Comment by bratpiorka** [comment]
btw - I added a separate JIRA task for extended test for HOST memspace (UMFW-246)

File: src/memspaces/memspace_host.c (hunk 7 above)
Comment (kswiecicki) [reply src/memspaces/memspace_host.c]: Done.

File: test/CMakeLists.txt [hunk 8]
```diff
@@ -50,7 +50,7 @@ function(add_umf_test)
     set_tests_properties(${TEST_NAME} PROPERTIES LABELS "umf")
 
     if(WIN32)
-        set_property(TEST ${TEST_NAME} PROPERTY ENVIRONMENT_MODIFICATION 
+        set_property(TEST ${TEST_NAME} PROPERTY ENVIRONMENT_MODIFICATION
```
[the remark this answers is not in the document]
Comment (kswiecicki) [reply test/CMakeLists.txt:53]: I'm not sure if its capable of working in `CMakeLists.txt` files.

File: src/memspaces/memspace_host.c [hunk 9]
```diff
@@ -0,0 +1,98 @@
...
+        // TODO: What would be an approrpiate err?
+        return UMF_RESULT_ERROR_UNKNOWN;
+    }
+
+    if (hwloc_topology_load(topology)) {
+        umf_ret = UMF_RESULT_ERROR_UNKNOWN;
+        goto err_topology_destroy;
+    }
+
+    // Shouldn't return -1, as 'HWLOC_OBJ_NUMANODE' deosn't appear to be an
```
Comment (igchor) [inline src/memspaces/memspace_host.c:37]: pehaps add and assert for that?

File: src/memspaces/memspace_host.c [hunk 10]
```diff
@@ -0,0 +1,98 @@
...
+}
+
+static umf_memspace_handle_t UMF_MEMSPACE_HOST = NULL;
+
+umf_result_t umfMemspaceHostGet(umf_memspace_handle_t *hMemspace) {
+    if (!hMemspace) {
+        return UMF_RESULT_ERROR_INVALID_ARGUMENT;
+    }
+
+    if (!UMF_MEMSPACE_HOST) {
```
Comment (igchor) [inline src/memspaces/memspace_host.c:82]: This should be under pthread_once or in library ctor - we want to allow calling umfMemspaceHostGet simultaneously from multiple threads.

File: src/memspaces/memspace_host.c (hunk 9 above)
Comment (kswiecicki) [reply src/memspaces/memspace_host.c]: Done, but I've set the `UMF_RESULT_ERROR_UNKNOWN` return value instead of assert.

File: src/memspaces/memspace_host.c (hunk 10 above)
Comment (kswiecicki) [reply src/memspaces/memspace_host.c]: Done.

**Review by igchor — APPROVED** [review]

File: src/CMakeLists.txt [hunk 11]
```diff
@@ -47,14 +47,20 @@ set(UMF_SOURCES_MACOSX
...
         provider/provider_os_memory.c
         provider/provider_os_memory_linux.c
         memory_targets/memory_target_numa.c
-        memspaces/memspace_numa.c)
+        memspaces/memspace_numa.c
+        memspaces/memspace_host.c)
     if(LINUX)
         set(UMF_LIBS ${UMF_LIBS} hwloc)
+        # Currently, OS provider is only built for Linux
+        set(UMF_COMPILE_DEFINITIONS ${UMF_COMPILE_DEFINITIONS} "UMF_BUILD_OS_MEMORY_PROVIDER")
```
Comment (bratpiorka) [inline src/CMakeLists.txt:63]: in "benchmark\CMakeLists.txt" there is a similar definition:

if (UMF_BUILD_OS_MEMORY_PROVIDER)
	target_compile_definitions(ubench PRIVATE UMF_BUILD_OS_MEMORY_PROVIDER=1)
endif()

could you add this definition the same way? In the future I think we should make this defines PUBLIC and cleanup cmake files but in the separate PR. You could even add a TODO for that

File: src/libumf.def [hunk 12]
```diff
@@ -16,6 +16,7 @@ EXPORTS
     umfMemoryProviderAllocationMerge
     umfMemoryProviderAllocationSplit
     umfMemoryProviderCreate
+    umfMemoryProviderCreateFromMemspace
```
Comment (bratpiorka) [inline src/libumf.def:19]: this is one of the reasons why I would love this file to be generated :)

**Review by bratpiorka — CHANGES_REQUESTED** [review]

File: include/umf/memspace.h [hunk 13]
```diff
@@ -44,6 +44,12 @@ umfMemoryProviderCreateFromMemspace(umf_memspace_handle_t hMemspace,
                                     umf_memspace_policy_handle_t hPolicy,
                                     umf_memory_provider_handle_t *hProvider);
 
+///
+/// \brief Retrieves predefined host memspace.
+/// \return UMF_RESULT_SUCCESS on success or appropriate error code on failure.
+///
+umf_result_t umfMemspaceHostGet(umf_memspace_handle_t *hMemspace);
```
Comment (vinser52) [inline include/umf/memspace.h:51]: Are we using getter function instead of predefined global memspace variable for the same reason (to be able to build UMF as a static lib) we replaced global OPS structures for pools/providers?

File: include/umf/memspace.h [hunk 14]
```diff
@@ -44,6 +44,12 @@ umfMemoryProviderCreateFromMemspace(umf_memspace_handle_t hMemspace,
                                     umf_memspace_policy_handle_t hPolicy,
                                     umf_memory_provider_handle_t *hProvider);
 
+///
+/// \brief Retrieves predefined host memspace.
+/// \return UMF_RESULT_SUCCESS on success or appropriate error code on failure.
```
[the remark this answers is not in the document]
Comment (kswiecicki) [reply include/umf/memspace.h:50]: Done.

File: src/CMakeLists.txt (hunk 11 above)
Comment (kswiecicki) [reply src/CMakeLists.txt]: Done.

File: include/umf/memspace.h [hunk 15]
```diff
@@ -44,6 +44,13 @@ umfMemoryProviderCreateFromMemspace(umf_memspace_handle_t hMemspace,
                                     umf_memspace_policy_handle_t hPolicy,
                                     umf_memory_provider_handle_t *hProvider);
 
+///
+/// \brief Retrieves predefined host memspace.
```
Comment (vinser52) [inline include/umf/memspace.h:48]: What does "host" memspace mean? Is it something similar to the default kind in Memkind? Anyway, I think we need a more descriptive name.

Comment (kswiecicki) [reply include/umf/memspace.h:48]: It's a memspace backed by numa nodes. Perhaps `UMF_MEMSPACE_NUMA` would be better?

File: include/umf/memspace.h (hunk 13 above)
Comment (kswiecicki) [reply include/umf/memspace.h:52]: Yeah, @igchor mentioned that it would be for the best if host memspace followed the same approach as with ops.

File: src/CMakeLists.txt [hunk 16]
```diff
@@ -47,14 +47,21 @@ set(UMF_SOURCES_MACOSX
     libumf_linux.c
 )
 
+# Compile definitions for UMF library
+# TODO: Cleanup the compile definitions across all the Cmake files
+set(UMF_COMPILE_DEFINITIONS "")
```
[the remark this answers is not in the document]
Comment (kswiecicki) [reply src/CMakeLists.txt:52]: I can't use `target_compile_definitions()` on a target that was not yet created.

**Comment by kswiecicki** [comment]
I've seen the changes in https://github.com/oneapi-src/unified-memory-framework/pull/229/ and adopted the same approach to host memspace.

File: include/umf/memspace.h (hunk 14 above)
[the remark this answers is not in the document]
Comment (kswiecicki) [reply include/umf/memspace.h]: Actually, I've changed the signature of `umfMemspaceHostGet` to no longer accept a param.

File: include/umf/memspace.h (hunk 15 above)
Comment (bratpiorka) [reply include/umf/memspace.h:48]: I think we should call it the same as in memkind - "defaut". In memkind, all NUMA nodes available for allocation (so also, for example, PMEM nodes set to KMEM_DAX mode) created a "default" namespace. In addition, NUMA nodes that contain the CPU (= DRAM memory) created the "regular" kind.

**Review by bratpiorka — APPROVED** [review]

File: test/memspace_host.cpp [hunk 17]
```diff
@@ -0,0 +1,74 @@
...
+
+#define SIZE_4K (4096)
+
+struct predefinedMemspaceTest : test {
+
+    void SetUp() override {
+        ::test::SetUp();
+
+        hMemspace = umfMemspaceHostGet();
+        ASSERT_NE(hMemspace, nullptr);
```
Comment (lukaszstolarczuk) [inline test/memspace_host.cpp:19]: perhaps include `test_helpers.h` and use our `UT_ASSERT*` macros

File: src/memspaces/memspace_host.c
Comment (lukaszstolarczuk) [inline src/memspaces/memspace_host.c:1]: could you please provide at lest a little description in the commit message (to the commit "Add predefined HOST memspace")

File: test/memspace_host.cpp (hunk 17 above)
Comment (kswiecicki) [reply test/memspace_host.cpp]: Done.

File: src/memspaces/memspace_host.c
Comment (kswiecicki) [reply src/memspaces/memspace_host.c:1]: I've added a more descriptive commit message.

Comment (lukaszstolarczuk) [reply src/memspaces/memspace_host.c:1]: thx!

**Review by lukaszstolarczuk — COMMENTED** [review]
👍

File: include/umf/memspace.h (hunk 15 above)
Comment (kswiecicki) [reply include/umf/memspace.h]: I've changed the memspace name to `default`.

File: README.md [hunk 18]
```diff
@@ -161,6 +161,18 @@ The `UMF_BUILD_LIBUMF_POOL_SCALABLE` option has to be turned `ON` to build this
 2) Required packages:
    - libtbb-dev (libraries: libtbbmalloc.so.2)
 
+### Memspaces (Linux-only)
+
+TODO: Add general information about memspaces.
+
+#### Default memspace
+
+Memspace backed by NUMA node discovery. Can be retrieved using umfMemspaceDefaultGet.
```
Comment (vinser52) [inline README.md:170]: Can we describe more precisely what “dafault” means (e.g. all available NUMA nodes)? Or we intentionaly do not specify this for some abstraction purposes.

File: test/memspace_default.cpp [hunk 19]
```diff
@@ -0,0 +1,75 @@
...
+
+    umf_memory_provider_handle_t provider = nullptr;
+};
+
+TEST_F(test, memspaceGet) {
+    umf_memspace_handle_t hMemspace = umfMemspaceDefaultGet();
+    UT_ASSERTne(hMemspace, nullptr);
+}
+
+TEST_F(predefinedMemspaceTest, providerFromMemspace) {
```
Comment (vinser52) [inline test/memspace_default.cpp:51]: This test looks like a duplication. The same flow is done and checked in the `SetUp` function.

File: README.md (hunk 18 above)
Comment (kswiecicki) [reply README.md]: I've changed the brief about `default` memspace. The info that it's backed by all available NUMA nodes is now transparent.

File: include/umf/memspace.h [hunk 20]
```diff
@@ -44,6 +44,12 @@ umfMemoryProviderCreateFromMemspace(umf_memspace_handle_t hMemspace,
                                     umf_memspace_policy_handle_t hPolicy,
                                     umf_memory_provider_handle_t *hProvider);
 
+///
+/// \brief Retrieves predefined default memspace.
+/// \return default memspace handle on success or NULL on failure.
+///
+umf_memspace_handle_t umfMemspaceDefaultGet(void);
```
Comment (vinser52) [inline include/umf/memspace.h:51]: An alternative API could be a single entry function which accept some enum to specify which predefined memspace to return. For example:
```
enum predefined_memspaces {
    DEFAULT,
    ...
};

umf_memspace_handle_t umfPredefinedMemspaceGet(enum predefined_memspaces) {
    ...
}
```

@igchor, @bratpiorka, @pbalcer what do you think?

File: src/memspaces/memspace_default.c [hunk 21]
```diff
@@ -0,0 +1,91 @@
...
+    umfMemspaceDestroy(UMF_MEMSPACE_DEFAULT);
+    UMF_MEMSPACE_DEFAULT = NULL;
+}
+
+static void umfMemspaceDefaultInit(void) {
+    umf_result_t ret = umfMemspaceDefaultCreate(&UMF_MEMSPACE_DEFAULT);
+    assert(ret == UMF_RESULT_SUCCESS);
+    (void)ret;
+
+    atexit(umfMemspaceDefaultDestroy);
```
[the remark this answers is not in the document]
Comment (igchor) [reply src/memspaces/memspace_default.c:85]: The problem is that there's no library destructor on windows when we build as a static library. I think we can do the same as for base_alloc: https://github.com/oneapi-src/unified-memory-framework/blob/main/src/base_alloc/base_alloc_global.c#L63 We use atexit when building as static lib on windows and call umf_ba_destroy_global in library dtor otherwise.

File: include/umf/memspace.h (hunk 20 above)
Comment (igchor) [reply include/umf/memspace.h:51]: Yes, this is also possible. I don't really have a preference here.

Comment (bratpiorka) [reply include/umf/memspace.h:51]: Currently, we use umfXXXPool/ProviderOps() for pools and providers. We should be consistent with that or change this everywhere. But please note that the enum is more problematic when you want to build this list at cmake configuration (enable/disable some compontents).

Comment (vinser52) [reply include/umf/memspace.h:51]: > Currently, we use umfXXXPool/ProviderOps() for pools and providers. We should be consistent with that or change this everywhere.

I do not think the same rules apply everywhere.

> But please note that the enum is more problematic when you want to build this list at cmake configuration (enable/disable some components).

Ok, it makes sense if we want to configure the available predefined memspaces at build-time.

File: src/memspaces/memspace_default.c (hunk 21 above)
[the remark this answers is not in the document]
Comment (vinser52) [reply src/memspaces/memspace_default.c:85]: In general, what Igor suggested makes sense if we want to support dynamic and static versions of UMF. 
But eventually building UMF core as a static library does not make much sense - tracking would not work (as expected).

So I agree with what @igchor suggested.

[the remark this answers is not in the document]
Comment (kswiecicki) [reply src/memspaces/memspace_default.c]: Done. I've followed the example that @igchor gave. I've removed the `atexit` altogether since DEFAULT memspace depends on OS provider that is unsupported on Windows for now.

**Review by PatKamin — APPROVED** [review]

**Review by ldorau — CHANGES_REQUESTED** [review]
Please rebase on top of the current main branch where the proxy library is present.

**Review by ldorau — APPROVED** [review]
